File: Spring/projects/kimyounghan/spring-db2/spring-transaction/src/main/java/org/festilog/springtransaction/propagation/MemberRepository.java
package org.festilog.springtransaction.propagation;

import jakarta.persistence.EntityManager;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Slf4j
@Repository
@RequiredArgsConstructor
public class MemberRepository {

    private final EntityManager em;

//    @Transactional // JPA의 모든 데이터 변경은 트랜잭션 안에서 이루어져야한다.
    public void save(final Member member) {
        log.info("member 저장");
        em.persist(member);
    }

    public Optional<Member> findByUsername(final String username) {
        return em.createQuery("select m from Member m where m.username = :username", Member.class)
                .setParameter("username", username)
                .getResultList().stream().findAny();
    }
}

File: projects/kotlin/kt-security-example/src/main/kotlin/com/example/ktsecurityexample/member/repository/MemberRepository.kt
package com.example.ktsecurityexample.member.repository

import com.example.ktsecurityexample.member.entity.Member
import org.springframework.data.jpa.repository.JpaRepository

interface MemberRepository : JpaRepository<Member, Long> {
    fun findByLoginId(loginId: String): Member?
}
